azure/services/roleassignments/spec.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package roleassignments
+
+import (
+	"github.com/Azure/azure-sdk-for-go/profiles/2019-03-01/authorization/mgmt/authorization"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+)
+
+// RoleAssignmentSpec defines the specification for a role assignment.
+type RoleAssignmentSpec struct {
+	RoleAssignmentName string
+	ResourceGroup      string
+	MachineName        string
+	Name               string
+	ResourceType       string
+	SubscriptionID     string
+	PrincipalID        *string
+	RoleDefinitionID   string
+	Scope              string
+}
+
+// ResourceName returns the name of the role assignment.
+func (s *RoleAssignmentSpec) ResourceName() string {
+	return s.RoleAssignmentName
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *RoleAssignmentSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName is a no-op for role assignments.
+func (s *RoleAssignmentSpec) OwnerResourceName() string {
+	return ""
```

Comment:
is role assignment not a sub resource of the VM/VMSS ? what if this returned the scope?

Reply:
>is role assignment not a sub resource of the VM/VMSS ? what if this returned the scope?
If it returns the scope -- it basically returns -- `fmt.Sprintf("/subscriptions/%s/", subscriptionID)`

Reply:
I would vote for returning the scope since it would really simplify the code and avoid having to cast parameters in the client, especially since we're not using `OwnerResourceName` for anything. Technically the scope isn't an "owning resource" but it does define what parent resource the role assignment applies to, and it's part of the role assignment resource ID, just like a parent resource: `PUT https://management.azure.com/scope/providers/Microsoft.Authorization/roleAssignments/roleAssignmentName?api-version=2015-07-01`

https://docs.microsoft.com/en-us/rest/api/authorization/role-assignments/create#create-role-assignment


We could even consider renaming `OwnerResourceName` to something that includes both "scope" and "owner resource name" if that improves code readability. At the very least we would add a code comment on this func to explain that it returns the scope for role assignments.

Reply:
found some more official docs on this: https://docs.microsoft.com/en-us/azure/role-based-access-control/scope-overview#scope-and-arm-templates:

> A role assignment is a special type in Azure Resource Manager called an extension resource. An extension resource is a resource that adds to another resource's capabilities. They always exist as an extension (like a child) of another resource. For example, a role assignment at subscription scope is an extension resource of the subscription. The name of a role assignment is always the name of the resource you are extending plus /Microsoft.Authorization/roleAssignments/{roleAssignmentId}

Reply:
Sure. Returning scope from `OwnerResourceName` makes sense to me. I will make the changes accordingly.